File: java/5-longest-palindromic-substring.java
class Solution {
    String ans;

    public String longestPalindrome(String s) {
        ans = s.substring(0, 1);
        for (int i = 0; i < s.length() - 1; i++) {
            findPalindrome(i, i, s);// odd string
            findPalindrome(i, i + 1, s);// even string
        }
        return ans;
    }

    private void findPalindrome(int l, int r, String s) {
        while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            l--;
            r++;
        }
        if (ans.length() < r - 1 - l) {
            ans = s.substring(l + 1, r);
        }
    }
}

File: hao/c++/5-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int maxLength = 0;
        int length = s.length();
        int j = 1;
        int tempLength = 1;
        int bestStart = 0;
        for (int i = 0; i < length; i++) {
            // Length: odd
            j = 1;
            while (true) {
                if (i-j < 0 || i+j > length-1 || s[i-j] != s[i+j]) {
                    break;
                }
                j++;
            }
            tempLength = 2*j-1;
            if (tempLength > maxLength) {
                maxLength = tempLength;
                bestStart = i-j+1;
            }
            
            // Length: Even
            if (s[i] != s[i+1] || i == length - 1) {
                continue;
            }
            j = 1;
            while (true) {
                
                if (i-j < 0 || i+j+1 > length-1 || s[i-j] != s[i+j+1]) {
                    break;
                }
                j++;
            }
            tempLength = 2*j;
            if (tempLength > maxLength) {
                maxLength = tempLength;
                bestStart = i-j+1;
            }
        }
        
        // Output
        string t = s.substr(bestStart, maxLength);
        return t;
    }
    
};